File: web/src/auth/TwoFactor.js
// Copyright 2023 The Casdoor Authors. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import React, {useState} from "react";
import i18next from "i18next";
import {Button, Input} from "antd";
import * as Setting from "../Setting";
import {twoFactorAuthTotp, twoFactorRecoverTotp} from "../backend/TwoFactorAuthBackend";

export const NextTwoFactor = "nextTwoFactor";

export function VerityTotp({onSuccess, onFail}) {
  const [loading, setLoading] = useState(false);
  const [currentStep, setCurrentStep] = useState(0);
  const [passcode, setPasscode] = useState("");
  const [recoveryCode, setRecoveryCode] = useState("");
  const verity = () => {
    if (passcode.length === 0) {
      Setting.showMessage("error", i18next.t("two-factor:Enter passcode"));
      return;
    }

    setLoading(true);
    twoFactorAuthTotp({passcode}).then((res) => {
      if (res.status === "ok") {
        onSuccess();
      } else {
        onFail(res.msg);
      }
    }).catch((reason) => {
      onFail(reason.message);
    }).finally(() => {
      setLoading(false);
    });
  };

  const recover = () => {
    if (recoveryCode.length === 0) {
      Setting.showMessage("error", i18next.t("two-factor:Enter recovery code"));
      return;
    }
    setLoading(true);
    twoFactorRecoverTotp({recoveryCode}).then(res => {
      if (res.status === "ok") {
        onSuccess();
      } else {
        onFail(res.msg);
      }
    }).catch((reason) => {
      onFail(reason.message);
    }).finally(() => {
      setLoading(false);
    });
  };

  switch (currentStep) {
  case 0:
    return (
      <div style={{width: 300}}>
        <div style={{marginBottom: 24, textAlign: "center", fontSize: "28px"}}>{i18next.t(
          "two-factor:Two-factor authentication")}</div>
        <div style={{marginBottom: 24}}>{i18next.t(
          "two-factor:Two-factor authentication description")}</div>
        <Input placeholder={i18next.t("two-factor:Passcode")}
          style={{marginBottom: 24}} type={"passcode"} size={"large"}
          onChange={event => setPasscode(event.target.value)} />
        <Button style={{width: "100%"}} size={"large"} loading={loading}
          type={"primary"} onClick={() => {
            verity();
          }}>{i18next.t("two-factor:Verity")}</Button>
        <span style={{float: "right"}}>
          {i18next.t("two-factor:Have problems?")}
          <a onClick={() => {
            setCurrentStep(1);
          }}>
            {i18next.t("two-factor:Use a recovery code")}
          </a>
        </span>
      </div>
    );
  case 1:
    return (
      <div style={{width: 300}}>
        <div style={{marginBottom: 24, textAlign: "center", fontSize: "28px"}}>{i18next.t(
          "two-factor:Two-factor recover")}</div>
        <div style={{marginBottom: 24}}>{i18next.t(
          "two-factor:Two-factor recover description")}</div>
        <Input placeholder={i18next.t("two-factor:Recovery code")}
          style={{marginBottom: 24}} type={"passcode"} size={"large"}
          onChange={event => setRecoveryCode(event.target.value)} />
        <Button style={{width: "100%"}} size={"large"} loading={loading}
          type={"primary"} onClick={() => {
            recover();
          }}>{i18next.t("two-factor:Verity")}</Button>
      </div>
    );
  default:
    return null;
  }
}

File: controllers/two_factor.go
// Copyright 2023 The Casdoor Authors. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package controllers

import (
	"net/http"

	"github.com/beego/beego"
	"github.com/casdoor/casdoor/object"
	"github.com/casdoor/casdoor/util"
)

// TwoFactorSetupInitiate
// @Title TwoFactorSetupInitiate
// @Tag Two-Factor API
// @Description setup totp
// @param userId	form	string	true	" "<owner>/<name>" of user"
// @Success 200 {object}   The Response object
// @router /mfa/setup/initiate [post]
func (c *ApiController) TwoFactorSetupInitiate() {
	owner := c.Ctx.Request.Form.Get("owner")
	name := c.Ctx.Request.Form.Get("name")
	authType := c.Ctx.Request.Form.Get("type")
	userId := util.GetId(owner, name)

	if len(userId) == 0 {
		c.ResponseError(http.StatusText(http.StatusBadRequest))
		return
	}

	twoFactorUtil := object.GetTwoFactorUtil(authType, nil)
	if twoFactorUtil == nil {
		c.ResponseError("Invalid auth type")
	}
	user := object.GetUser(userId)
	if user == nil {
		c.ResponseError("User doesn't exist")
		return
	}

	issuer := beego.AppConfig.String("appname")
	accountName := user.GetId()

	twoFactorProps, err := twoFactorUtil.Initiate(c.Ctx, issuer, accountName)
	if err != nil {
		c.ResponseError(err.Error())
		return
	}

	resp := twoFactorProps
	c.ResponseOk(resp)
}

// TwoFactorSetupVerify
// @Title TwoFactorSetupVerify
// @Tag Two-Factor API
// @Description setup verify totp
// @param	secret		form	string	true	"totp secret"
// @param	passcode	form 	string 	true	"totp passcode"
// @Success 200 {object}  Response object
// @router /mfa/setup/totp/verify [post]
func (c *ApiController) TwoFactorSetupVerify() {
	authType := c.Ctx.Request.Form.Get("type")
	passcode := c.Ctx.Request.Form.Get("passcode")

	if authType == "" || passcode == "" {
		c.ResponseError("missing auth type or passcode")
		return
	}
	twoFactorUtil := object.GetTwoFactorUtil(authType, nil)

	err := twoFactorUtil.SetupVerify(c.Ctx, passcode)
	if err != nil {
		c.ResponseError(err.Error())
	} else {
		c.ResponseOk(http.StatusText(http.StatusOK))
	}
}

// TwoFactorSetupEnable
// @Title TwoFactorSetupEnable
// @Tag Two-Factor API
// @Description enable totp
// @param	userId		form	string	true	"Id of user"
// @param  	secret		form	string	true	"totp secret"
// @Success 200 {object}  Response object
// @router /mfa/setup/enable [post]
func (c *ApiController) TwoFactorSetupEnable() {
	owner := c.Ctx.Request.Form.Get("owner")
	name := c.Ctx.Request.Form.Get("name")
	authType := c.Ctx.Request.Form.Get("type")

	user := object.GetUser(util.GetId(owner, name))
	if user == nil {
		c.ResponseError("User doesn't exist")
		return
	}

	twoFactor := object.GetTwoFactorUtil(authType, nil)
	err := twoFactor.Enable(c.Ctx, user)
	if err != nil {
		c.ResponseError(err.Error())
		return
	}

	c.ResponseOk(http.StatusText(http.StatusOK))
}

// TwoFactorDelete
// @Title TwoFactorDelete
// @Tag Two-Factor API
// @Description: Remove Totp
// @param	userId	form	string	true	"Id of user"
// @Success 200 {object}  Response object
// @router /mfa/ [delete]
func (c *ApiController) TwoFactorDelete() {
	id := c.Ctx.Request.Form.Get("id")
	owner := c.Ctx.Request.Form.Get("owner")
	name := c.Ctx.Request.Form.Get("name")
	userId := util.GetId(owner, name)

	user := object.GetUser(userId)
	if user == nil {
		c.ResponseError("User doesn't exist")
		return
	}

	twoFactorProps := user.TwoFactorAuth[:0]
	i := 0
	for _, twoFactorProp := range twoFactorProps {
		if twoFactorProp.Id != id {
			twoFactorProps[i] = twoFactorProp
			i++
		}
	}
	user.TwoFactorAuth = twoFactorProps
	object.UpdateUser(userId, user, []string{"two_factor_auth"}, user.IsAdminUser())
	c.ResponseOk(user.TwoFactorAuth)
}
